### DISCUSSION TITLE
Does zitadel support custom roles and permission in tenant organization?

### DISCUSSION THREAD

**Discussion by stiwari99** [zitadel/zitadel#6927] [Q&A] [answered]
I have this scenario, wondering if zitadel will be able to solve this

As a saas provider we have list of permissions
**Permission list**
1. read_invoice
2. manage_invoice
3. delete_invoice
4. read_price
5. mange_price
6. delete_price

**Service defined Roles**
There are also pre-defined roles in service which are collection of permissions
1. org_admin: Permissions - (read_invoice, manage_invoice, delete_invoice, read_price, manage_price, delete_price)
3. employee: Permissions - (read_invoice, read_price)

Say our customers are B2B, which means they are going to manage their own roles. Each B2B customer is an organization
Org1 (Pepsi)
Org2 (Coke)

Now by default both organizations should be able to use system defined roles (i.e org_admin and employee) and they should be able to create new custom roles as per their needs with permission defined by system

e.g 

- Org 1 (Pepsi)
   - List of Roles
      - org_admin
      - employee
      - manager (**custom role** with permission : read_invoice, manage_invoice, read_price, manage_price)
      - sub_manager (**custom role** with permission: read_invoice, manage_invoice, read_price)

- Org 2 (Pepsi)
   - List of Roles
      - org_admin
      - employee
      - staff_member (**custom role** with permission : read_invoice, read_price, manage_price)
      - floor_manager (**custom role** with permission: read_invoice, manage_invoice, read_price)
   
As we can see above org 1 and org 2 has service defined roles but they have also created custom roles as per their needs. They should be only able to create custom roles based on permissions which are defined by service.

Orgs should be able to assign any of above roles to their users.

Is this something can be done using zitadel

**Comment by mffap** [comment, marked as answer]
I assume the custom roles should only be available for that specific customer? Meaning Org 1 would see sub_manager but not floor_manager?

At the moment, I think it is hard to make this usecase work. You could give a (service) user or Org 1 or Org 2 the manager role ORG_PROJECT_PERMISSION_EDITOR for the project that holds the roles, but that would allow them to change everything about the roles and not only their own. Furthermore creating high-level roles based on more fine-grained permissions is currently not available in ZITADEL.  Maybe have a look at a specific tool like https://warrant.dev/ for access control and it's self-service.

**Reply by stiwari99** [reply]
Yes, that is correct, that orgs should not be allowed to view each other roles.

I think I was expecting that as of now zitadel does not support this use case.

Thanks for your answer and suggestion regarding warrant.dev. I will check this

**Reply by mario-jerkovic** [reply]
@bigbang4u2 What was your decision in the end? Did you stick with ZITADEL, move to Warrant, or choose something else? 🙂

**Reply by stiwari99** [reply]
Currently, our authentication and authorization system is built in-house, addressing various scenarios, including multi-tenancy. After conducting a proof of concept (POC) with Zitadel, we opted to integrate it into our solution due to its expanded functionality compared to our existing setup. While Zitadel doesn't directly manage custom roles created by tenants, we've devised a workaround. We create only standard fixed roles in Zitadel, such as "employee," "customer," and "accountant." Custom role management is handled within our application.

Here's an overview of our usage of Zitadel for authentication:

1. Establish fixed roles in Zitadel (e.g., employee, customer, accountant) to simplify role assignments for tenants when adding users or during self-registration.
2. Implement a bi-directional user sync between our application and Zitadel.
3. Allow users to set custom roles within our application, ensuring that these roles are synchronized with Zitadel's user metadata. This synchronization enables us to retrieve roles directly from the token during token introspection.
